### PR TITLE
Add 2 fuzzers for nu-path, nu-parser

### PR DESCRIPTION
**Pull request by tokatoka**
# Description

This PR adds a fuzzer for the nu-path and the nu-parser crate.
Now you can go to `crates/nu-path/fuzz`/`crates/nu-parser/fuzz` and run `cargo fuzz` to
find crashes.
https://github.com/nushell/nushell/issues/10365 and #9417 was found by
this
# User-Facing Changes

it doesn't affect user experience

### CODE REVIEW

**Review by sholderbach — COMMENTED** [review]
~~Please keep the `rust-toolchain.toml`, we use it to ensure the Rust version used by developers is compatible with the packaging "vendors".~~

~~Not sure about what is up with `crates/nu-path/src/util.rs`? Is this unlinked in the module tree?~~

If we land this into the tree it would be great to have a readme pointing to relevant documentation.

**Comment by tokatoka** [comment]
> Please keep the rust-toolchain.toml, we use it to ensure the Rust version used by developers is compatible with the packaging "vendors".

> Not sure about what is up with crates/nu-path/src/util.rs? Is this unlinked in the module tree?

yes I mistakenly pushed other changes :D

**Comment by sholderbach** [comment]
No problem! Thanks for sharing your work, this sounds helpful to have, great that you already found a problem through that.

**Comment by fdncred** [comment]
I think this is a great idea to introduce fuzzing. nu-path is a good place to start. Thanks!

**Comment by tokatoka** [comment]
I added the README.md now. This pr is ready

**Comment by fdncred** [comment]
`gather_seeds.sh` makes this look like it only works on *nix platforms. Is there no way to make this work in Windows too?

File: crates/nu-path/fuzz/README.md [hunk 1]
```diff
@@ -0,0 +1,8 @@
+# Fuzzer for `nu-parser`
```
Comment (sholderbach) [inline crates/nu-path/fuzz/README.md:1]:
```suggestion
# Fuzzer for `nu-path`
```

File: crates/nu-parser/fuzz/gather_seeds.sh [hunk 2]
```diff
@@ -0,0 +1,10 @@
+#!/bin/bash
+
+# Check if 'seeds' directory exists. If not, create one.
+if [ ! -d "seeds" ]; then
+    mkdir seeds
+fi
+
+# Find all "*.nu" files from '../..' excluding the 'seeds' directory and copy them into 'seeds' directory.
+find ../../.. -type f -name "*.nu" -exec cp {} seeds/ \;
+
```
Comment (sholderbach) [inline crates/nu-parser/fuzz/gather_seeds.sh:10]: don't know if `/bin/bash` as a locatoin is guaranteed :P

Should this step be mentioned in the Readme?

If you want a larger corpus of data, one could scrape the https://github.com/nushell/nu_scripts repo. Don't know where there are diminishing returns with regards to entropy.

**Review by sholderbach — COMMENTED** [review]
Haven't played with `cargo-fuzz` so far but the targets look good to me. 

Will give it a try later :)

File: crates/nu-parser/fuzz/gather_seeds.sh (hunk 2 above)
Comment (tokatoka) [reply crates/nu-parser/fuzz/gather_seeds.sh]: > don't know if /bin/bash as a locatoin is guaranteed :P

Now I switched to nuscript.

> If you want a larger corpus of data, one could scrape the https://github.com/nushell/nu_scripts repo. Don't know where there are diminishing returns with regards to entropy.

Yes usually more and diverse sets of initial inputs is always good. but for now i will keep this as default, and the fuzzing-practitioner can change it as they want/

**Comment by tokatoka** [comment]
> gather_seeds.sh makes this look like it only works on *nix platforms. Is there no way to make this work in Windows too? Seems like the same thing could be done with a nushell script and nu gather_seeds.nu.

actually `cargo fuzz` says it works only on *nix platforms. but anyways i switched to use nu

File: crates/nu-parser/fuzz/.gitignore [hunk 3]
```diff
@@ -0,0 +1,4 @@
+target
+corpus
+artifacts
+coverage
```
Comment (sholderbach) [inline crates/nu-parser/fuzz/.gitignore:4]: With the instructions provided this needs more ignores:

```
        crates/nu-parser/fuzz/Cargo.lock
        crates/nu-parser/fuzz/out/
        crates/nu-parser/fuzz/seeds/
```

Comment (tokatoka) [reply crates/nu-parser/fuzz/.gitignore:4]: ok

**Comment by tokatoka** [comment]
> Already found something new that triggers an access out ouf bounds panic:

You'll find many :D

**Comment by anka-213** [comment]
I just found a file that triggers a rapid memory leak, about 4 GB in a minute
[oom-1dae5132c7066b2c686fe27d2c0f247df4d17dd1.txt](https://github.com/nushell/nushell/files/12666138/oom-1dae5132c7066b2c686fe27d2c0f247df4d17dd1.txt)

**Comment by tokatoka** [comment]
nice :wink:

**Pull request merged**